test/fuzz/transformation_push_id_through_variable_test.cpp

```diff
@@ -0,0 +1,389 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_push_id_through_variable.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "test/fuzz/fuzz_test_util.h"
+
+namespace spvtools {
+namespace fuzz {
+namespace {
+
+std::string reference_shader = R"(
+             OpCapability Shader
+        %1 = OpExtInstImport "GLSL.std.450"
+             OpMemoryModel Logical GLSL450
+             OpEntryPoint Fragment %4 "main" %92 %52 %53
+             OpExecutionMode %4 OriginUpperLeft
+             OpSource ESSL 310
+             OpDecorate %92 BuiltIn FragCoord
+        %2 = OpTypeVoid
+        %3 = OpTypeFunction %2
+        %6 = OpTypeInt 32 1
+        %7 = OpTypeFloat 32
+        %8 = OpTypeStruct %6 %7
+        %9 = OpTypePointer Function %8
+       %10 = OpTypeFunction %6 %9
+       %14 = OpConstant %6 0
+       %15 = OpTypePointer Function %6
+       %51 = OpTypePointer Private %6
+       %21 = OpConstant %6 2
+       %23 = OpConstant %6 1
+       %24 = OpConstant %7 1
+       %25 = OpTypePointer Function %7
+       %50 = OpTypePointer Private %7
+       %34 = OpTypeBool
+       %35 = OpConstantFalse %34
+       %60 = OpConstantNull %50
+       %61 = OpUndef %51
+       %52 = OpVariable %50 Private
+       %53 = OpVariable %51 Private
+       %80 = OpConstantComposite %8 %21 %24
+       %90 = OpTypeVector %7 4
+       %91 = OpTypePointer Input %90
+       %92 = OpVariable %91 Input
+       %93 = OpConstantComposite %90 %24 %24 %24 %24
+        %4 = OpFunction %2 None %3
+        %5 = OpLabel
+       %20 = OpVariable %9 Function
+       %27 = OpVariable %9 Function
+       %22 = OpAccessChain %15 %20 %14
+       %44 = OpCopyObject %9 %20
+       %26 = OpAccessChain %25 %20 %23
+       %29 = OpFunctionCall %6 %12 %27
+       %30 = OpAccessChain %15 %20 %14
+       %45 = OpCopyObject %15 %30
+       %81 = OpCopyObject %9 %27
+       %33 = OpAccessChain %15 %20 %14
+             OpSelectionMerge %37 None
+             OpBranchConditional %35 %36 %37
+       %36 = OpLabel
+       %38 = OpAccessChain %15 %20 %14
+       %40 = OpAccessChain %15 %20 %14
+       %43 = OpAccessChain %15 %20 %14
+       %82 = OpCopyObject %9 %27
+             OpBranch %37
+       %37 = OpLabel
+             OpReturn
+             OpFunctionEnd
+       %12 = OpFunction %6 None %10
+       %11 = OpFunctionParameter %9
+       %13 = OpLabel
+       %46 = OpCopyObject %9 %11
+       %16 = OpAccessChain %15 %11 %14
+       %95 = OpCopyObject %8 %80
+             OpReturnValue %21
+             OpFunctionEnd
+)";
+
+const auto env = SPV_ENV_UNIVERSAL_1_4;
+const auto consumer = nullptr;
+const auto context =
+    BuildModule(env, consumer, reference_shader, kFuzzAssembleOption);
+
+FactManager fact_manager;
+spvtools::ValidatorOptions validator_options;
+TransformationContext transformation_context(&fact_manager, validator_options);
```

Comment:
I am unsure about making these global.  I'm not sure how gtest works but I would not be surprised if the state of globals persists between individual tests.  We do not want to require the tests to run in a particular order.

I suggest you follow the style used by other tests, where each test creates its own IR context locally.  Alternatively you could look at other facilities that gtest might offer, such as test fixtures - they could be useful if you want to reduce redundancy between tests.

That said, I quite like each test to be pretty simple and self-contained, even if it ends up being a bit long.

@paulthomson your feedback here would be great.

Reply:
I agree with Ally here.

Reply:
Done.